driver-core/src/main/java/com/datastax/driver/core/CodecRegistry.java

```diff
@@ -0,0 +1,760 @@
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(javaType, cqlType);
+        }
+    }
+
+    private final CopyOnWriteArrayList<TypeCodec<?>> codecs;
+
+    private final ConcurrentMap<String, TypeCodec<?>> overrides;
+
+    private final LoadingCache<CacheKey, TypeCodec<?>> cache;
+
+    public CodecRegistry() {
+        this.codecs = new CopyOnWriteArrayList<TypeCodec<?>>(DEFAULT_CODECS);
+        this.overrides = new ConcurrentHashMap<String, TypeCodec<?>>();
+        this.cache = CacheBuilder.newBuilder()
+            .build(
+                new CacheLoader<CacheKey, TypeCodec<?>>() {
+                    public TypeCodec<?> load(CacheKey cacheKey) {
+                        return findCodec(cacheKey.cqlType, cacheKey.javaType);
+                    }
+                });
+    }
+
+    /**
+     * Register the given codec with this registry.
+     *
+     * @param codec The codec to add to the registry.
+     * @return this CodecRegistry (for method chaining).
+     */
+    public CodecRegistry register(TypeCodec<?> codec) {
+        return register(codec, false);
+    }
+
+    /**
+     * Register the given codec with this registry.
+     *
+     * @param codec The codec to add to the registry.
+     * @param includeDerivedCollectionCodecs If {@code true}, register not only the given codec
+     * but also collection codecs whose element types are handled by the given codec.
+     * @return this CodecRegistry (for method chaining).
+     */
+    public CodecRegistry register(TypeCodec<?> codec, boolean includeDerivedCollectionCodecs) {
+        return register(Collections.singleton(codec), includeDerivedCollectionCodecs);
+    }
+
+    /**
+     * Register the given codecs with this registry.
+     *
+     * @param codecs The codecs to add to the registry.
+     * @return this Builder (for method chaining).
+     */
+    public CodecRegistry register(TypeCodec<?>... codecs) {
+        return register(Arrays.asList(codecs), false);
+    }
+
+    /**
+     * Register the given codecs with this registry.
+     *
+     * @param codecs The codecs to add to the registry.
+     * @return this Builder (for method chaining).
+     */
+    public CodecRegistry register(Iterable<? extends TypeCodec<?>> codecs) {
+        return register(codecs, false);
+    }
+
+    /**
+     * Add the given codecs to the list of known codecs of this registry.
+     *
+     * @param codecs The codecs to add to the registry.
+     * @param includeDerivedCollectionCodecs If {@code true}, register not only the given codecs
+     * but also collection codecs whose element types are handled by the given codecs.
+     * @return this Builder (for method chaining).
+     */
+    public CodecRegistry register(Iterable<? extends TypeCodec<?>> codecs, boolean includeDerivedCollectionCodecs) {
+        for (TypeCodec<?> codec : codecs) {
+            if(includeDerivedCollectionCodecs){
```

Comment:
Shouldn't this apply to Tuples and UDTs somehow too?  (Not sure how we could do that deterministically)